File: src/main/kotlin/org/kite9/diagram/visualization/planarization/mgt/MGTPlanarization.java
package org.kite9.diagram.visualization.planarization.mgt;

import java.util.List;
import java.util.Set;

import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.edge.PlanarizationEdge;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.visualization.planarization.rhd.RHDPlanarization;

/**
 * Specific methods for GT-style planarization, in which the initial planarization is based on 
 * an ordered list of vertices.
 * 
 * @author moffatr
 * 
 */
public interface MGTPlanarization extends RHDPlanarization {

    public boolean isAdjacency(Edge edge);

    public boolean crosses(Edge edge, boolean above);

    public boolean crosses(float i1, float i2, boolean above);

    /**
	 * Introduces a new vertex into the ordering after point i.
	 */
	public void addVertexToOrder(int i, Vertex insert);
	
	public void removeVertexFromOrder(Vertex v);
	
    public int getVertexIndex(Vertex v);
    
    public List<Vertex> getVertexOrder();
    
    public void addEdge(PlanarizationEdge toAdd, boolean above, PlanarizationEdge outsideOf);
    
    /**
     * For a given vertex, returns edges leaving vertex above the line of the planarization
     * going forwards, in inside-most to outside-most order.
     */
    public List<PlanarizationEdge> getAboveForwardLinks(Vertex v);
	
	public List<PlanarizationEdge> getAboveBackwardLinks(Vertex v);
	
	public List<PlanarizationEdge> getBelowForwardLinks(Vertex v);
	
	public List<PlanarizationEdge> getBelowBackwardLinks(Vertex v);
	
	public Set<PlanarizationEdge> getAboveLineEdges();
	
	public Set<PlanarizationEdge> getBelowLineEdges();
	
	public PlanarizationEdge getFirstEdgeAfterPlanarizationLine(Vertex from, boolean forwardSet, boolean aboveSet);
}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/mgt/MGTPlanarization.kt
package org.kite9.diagram.visualization.planarization.mgt

import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.visualization.planarization.rhd.RHDPlanarization

/**
 * Specific methods for GT-style planarization, in which the initial planarization is based on
 * an ordered list of vertices.
 *
 * @author moffatr
 */
interface MGTPlanarization : RHDPlanarization {
    fun isAdjacency(edge: Edge): Boolean
    fun crosses(edge: Edge, above: Boolean): Boolean
    fun crosses(i1: Float, i2: Float, above: Boolean): Boolean

    /**
     * Introduces a new vertex into the ordering after point i.
     */
    fun addVertexToOrder(i: Int, insert: Vertex)
    fun getVertexIndex(v: Vertex): Int
    val vertexOrder: List<Vertex>
    fun addEdge(toAdd: PlanarizationEdge, above: Boolean, outsideOf: PlanarizationEdge?)

    /**
     * For a given vertex, returns edges leaving vertex above the line of the planarization
     * going forwards, in inside-most to outside-most order.
     */
    fun getAboveForwardLinks(v: Vertex): List<PlanarizationEdge>
    fun getAboveBackwardLinks(v: Vertex): List<PlanarizationEdge>
    fun getBelowForwardLinks(v: Vertex): List<PlanarizationEdge>
    fun getBelowBackwardLinks(v: Vertex): List<PlanarizationEdge>
    val aboveLineEdges: Set<PlanarizationEdge>
    val belowLineEdges: Set<PlanarizationEdge>
    fun getFirstEdgeAfterPlanarizationLine(from: Vertex, forwardSet: Boolean, aboveSet: Boolean): PlanarizationEdge
}